File: src/ios/SocialVk.m
//
//  SocialVk.m

#import "SocialVk.h"
#import <VKSdk/VKBundle.h>
#import "NSData+Base64.h"

static NSString *VK_AUTHORIZE_URL_STRING = @"vkauthorize://authorize";

@implementation SocialVk {
    CDVInvokedUrlCommand *savedCommand;
    void (^vkCallBackBlock)(NSString *, NSString *);
    BOOL inited;
    NSMutableDictionary *loginDetails;
}

@synthesize clientId;

- (void) initSocialVk:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    
    
    if(!inited) {
        NSString *appId = [[NSString alloc] initWithString:[command.arguments objectAtIndex:0]];
        [VKSdk initializeWithAppId:appId];
        [VKSdk.instance registerDelegate:self];
        VKSdk.instance.uiDelegate = self;
        [VKSdk wakeUpSession:@[VK_PER_OFFLINE] completeBlock:^(VKAuthorizationState state, NSError *err) {
            if(err) {
                NSLog(@"VK init error: %@", err);
            }
            if(state == VKAuthorizationAuthorized) {
                NSLog(@"VK user authorized with token %@", [VKSdk accessToken].accessToken);
            }
        }];
        NSLog(@"SocialVk Plugin initalized");
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(myOpenUrl:) name:CDVPluginHandleOpenURLNotification object:nil];
        inited = YES;
    }
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

-(UIViewController*)findViewController
{
    id vc = self.webView;
    do {
        vc = [vc nextResponder];
    } while([vc isKindOfClass:UIView.class]);
    return vc;
}

-(void)myOpenUrl:(NSNotification*)notification
{
    NSURL *url = notification.object;
    if(![url isKindOfClass:NSURL.class]) return;
    BOOL wasHandled = [VKSdk processOpenURL:url fromApplication:nil];
}

-(void) login:(CDVInvokedUrlCommand *)command
{
    NSArray *permissions = [command.arguments objectAtIndex:0];
    if(![VKSdk isLoggedIn]) {
        [self vkLoginWithPermissions:permissions andBlock:^(NSString *token, NSString *error) {
            if(token) {
                VKRequest *req = [VKRequest requestWithMethod:@"users.get" parameters:@{@"fields": @"id, nickname, first_name, last_name, sex, bdate, timezone, photo, photo_big, city, country"}];
                [req executeWithResultBlock:^(VKResponse *response) {
                    NSLog(@"User response %@", response);
                    
                    CDVPluginResult* pluginResult = nil;
                    loginDetails = [NSMutableDictionary new];
                    loginDetails[@"token"] = token;
                    if([response.json isKindOfClass:NSArray.class] && [(NSArray*)response.json count]>0 )
                        loginDetails[@"user"] = [response.json objectAtIndex:0];
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:loginDetails];
                    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
                } errorBlock:^(NSError *error) {
                    NSLog(@"Cant load user details");
                    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
                    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
                }];
            } else {
                NSLog(@"Cant login to VKontakte");
                CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }
        }];
    } else {
        if(loginDetails) {
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:loginDetails];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } else {
            VKAccessToken *token = [VKSdk accessToken];
            VKRequest *req = [VKRequest requestWithMethod:@"users.get" parameters:@{@"fields": @"sex,bdate,city,country,screen_name,photo_50,photo_200_orig"}];
            [req executeWithResultBlock:^(VKResponse *response) {
                NSLog(@"User response %@", response);
                CDVPluginResult* pluginResult = nil;
                loginDetails = [NSMutableDictionary new];
                loginDetails[@"token"] = token.accessToken;
                if([response.json isKindOfClass:NSArray.class] && [(NSArray*)response.json count]>0 )
                    loginDetails[@"user"] = [response.json objectAtIndex:0];
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:loginDetails];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            } errorBlock:^(NSError *error) {
                NSLog(@"Cant load user details");
                CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }];
        }
    }
}

-(void) share:(CDVInvokedUrlCommand*)command {
    NSBundle *vkb = [VKBundle vkLibraryResourcesBundle];
    NSLog(@"VK Bundle path %@", vkb.bundlePath);
    
    savedCommand = command;
    NSString *sourceURL = [command.arguments objectAtIndex:0];
    NSString* description = [command.arguments objectAtIndex:1];
    NSString* imageURL = [command.arguments objectAtIndex:2];
    //NSData *imageData = [NSData dataWithContentsOfURL:[NSURL URLWithString:imageURL]];

    if(![VKSdk isLoggedIn]) {
        [self vkLoginWithPermissions:nil andBlock:^(NSString *token, NSString *error) {
            CDVPluginResult* pluginResult = nil;
            if(token) {
                VKShareDialogController *sh = [VKShareDialogController new];
                sh.edgesForExtendedLayout = UIRectEdgeAll;
                sh.shareLink = [[VKShareLink alloc] initWithTitle:sourceURL link:[NSURL URLWithString:sourceURL]];
                sh.text =  description;
                //sh.uploadImages = @[[VKUploadImage uploadImageWithData:imageData andParams:nil]];
                UIViewController *vc = [self findViewController];
                [sh performSelector:@selector(presentIn:) withObject:vc afterDelay:2.f];
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            } else {
                NSLog(@"Cant login to VKontakte");
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }
        }];
    } else {
        VKShareDialogController *sh = [VKShareDialogController new];
        sh.edgesForExtendedLayout = UIRectEdgeAll;
        sh.shareLink = [[VKShareLink alloc] initWithTitle:sourceURL link:[NSURL URLWithString:sourceURL]];
        sh.text =  description;
        //sh.uploadImages = @[[VKUploadImage uploadImageWithData:imageData andParams:nil]];
        UIViewController *vc = [self findViewController];
        sh.dismissAutomatically = YES;
        [vc presentViewController:sh animated:YES completion:nil];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

-(void)vkLoginWithPermissions:(NSArray*)permissions andBlock:(void (^)(NSString *, NSString *))block
{
    vkCallBackBlock = [block copy];
    if(!permissions || permissions.count < 1)
    permissions = @[VK_PER_WALL, VK_PER_OFFLINE];
    //BOOL inApp = YES;
    //if([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:VK_AUTHORIZE_URL_STRING]])
    //    inApp = NO;
    //[VKSdk authorize:permissions revokeAccess:NO forceOAuth:NO inApp:inApp display:VK_DISPLAY_IOS];
    [VKSdk authorize:permissions];
}

-(void)logout:(CDVInvokedUrlCommand *)command
{
    [VKSdk forceLogout];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

#pragma mark - API methods

-(void)performRequest:(VKRequest*)request withCommand:(CDVInvokedUrlCommand*)command
{
    savedCommand = command;
    [request executeWithResultBlock:^(VKResponse *response) {
        savedCommand = nil;
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response.json];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    } errorBlock:^(NSError *error) {
        savedCommand = nil;
        NSLog(@"VK Error: %@", error);
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.description];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

-(void) users_get:(CDVInvokedUrlCommand*) command
{
    NSString* user_ids = [command.arguments objectAtIndex:0];
    NSString* fields = [command.arguments objectAtIndex:1];
    NSString* name_case = [command.arguments objectAtIndex:2];
    VKRequest *req = [[VKApi users] get:@{VK_API_USER_IDS: user_ids, VK_API_FIELDS: fields, VK_API_NAME_CASE: name_case}];
    [self performRequest:req withCommand:command];
}

-(void) users_search:(CDVInvokedUrlCommand*) command
{
    //query, sort, offset, count, fields, city, country, hometown, university_country, university, university_year, university_faculty, university_chair, sex, status, age_from, age_to, birth_day, birth_month, birth_year, online, has_photo, school_country, school_city, school_class, school, school_year, religion, interests, company, position, group_id, from_list
    id arg = [command.arguments objectAtIndex:0];
    NSDictionary *params;
    if([arg isKindOfClass:NSDictionary.class]) {
        params = arg;
    } else {
        params = @{VK_API_Q: arg};
    }
    VKRequest *req = [[VKApi users] search:params];
    [self performRequest:req withCommand:command];
}

- (void)users_isAppUser:(CDVInvokedUrlCommand*)command
{
    NSNumber *user_id = [command.arguments objectAtIndex:0];
    VKRequest *req;
    if(user_id && user_id.integerValue > 0)
        req = [[VKApi users] isAppUser:user_id.integerValue];
    else
        req = [[VKApi users] isAppUser];
    [self performRequest:req withCommand:command];
}

- (void)users_getSubscriptions:(CDVInvokedUrlCommand*)command
{
    NSNumber *user_id = [command.arguments objectAtIndex:0];
    NSNumber *extended = [command.arguments objectAtIndex:1];
    NSNumber *offset = [command.arguments objectAtIndex:2];
    NSNumber *count = [command.arguments objectAtIndex:3];
    NSString *fields = [command.arguments objectAtIndex:4];
    VKRequest *req = [[VKApi users] getSubscriptions:@{VK_API_USER_ID: user_id, VK_API_EXTENDED: extended, VK_API_OFFSET: offset, VK_API_COUNT: count, VK_API_FIELDS: fields}];
    [self performRequest:req withCommand:command];
}

- (void)users_getFollowers:(CDVInvokedUrlCommand*)command
{
    NSNumber *user_id = [command.arguments objectAtIndex:0];
    NSNumber *offset = [command.arguments objectAtIndex:1];
    NSNumber *count = [command.arguments objectAtIndex:2];
    NSString *fields = [command.arguments objectAtIndex:3];
    NSString *name_case = [command.arguments objectAtIndex:4];
    VKRequest *req = [[VKApi users] getFollowers:@{VK_API_USER_ID: user_id, VK_API_OFFSET: offset, VK_API_COUNT: count, VK_API_FIELDS: fields, VK_API_NAME_CASE: name_case}];
    [self performRequest:req withCommand:command];
}

- (void)wall_post:(CDVInvokedUrlCommand*)command
{
    id arg = [command.arguments objectAtIndex:0];
    NSDictionary *params;
    if([arg isKindOfClass:NSDictionary.class]) {
        params = arg;
    } else {
        params = @{VK_API_MESSAGE: arg};
    }
    VKRequest *req = [[VKApi wall] post:params];
    [self performRequest:req withCommand:command];
}

- (void)photos_getUploadServer:(CDVInvokedUrlCommand*)command
{
    NSNumber *album_id = [command.arguments objectAtIndex:0];
    NSNumber *group_id = [command.arguments objectAtIndex:1];
    VKRequest *req;
    if(group_id && group_id.integerValue > 0) {
        req = [[VKApi photos] getUploadServer:album_id.integerValue andGroupId:group_id.integerValue];
    } else {
        req = [[VKApi photos] getUploadServer:album_id.integerValue];
    }
    [self performRequest:req withCommand:command];
}

- (void)photos_getWallUploadServer:(CDVInvokedUrlCommand*)command
{
    NSNumber *group_id = [command.arguments objectAtIndex:0];
    VKRequest *req = [[VKApi photos] getWallUploadServer:group_id.integerValue];
    [self performRequest:req withCommand:command];
}

- (void)photos_saveWallPhoto:(CDVInvokedUrlCommand*)command
{
    NSString *imageBase64 = [command.arguments objectAtIndex:0];
    NSNumber *user_id = [command.arguments objectAtIndex:1];
    NSNumber *group_id = [command.arguments objectAtIndex:2];
    UIImage *image = [[UIImage alloc] initWithData:[NSData dataWithBase64EncodedString:imageBase64]];
    VKRequest *req = [VKApi uploadWallPhotoRequest:image parameters:[VKImageParameters pngImage] userId:user_id.integerValue groupId:group_id.integerValue];
    [self performRequest:req withCommand:command];
}

- (void)photos_save:(CDVInvokedUrlCommand*)command
{
    NSString *imageBase64 = [command.arguments objectAtIndex:0];
    NSNumber *album_id = [command.arguments objectAtIndex:1];
    NSNumber *group_id = [command.arguments objectAtIndex:2];
    UIImage *image = [[UIImage alloc] initWithData:[NSData dataWithBase64EncodedString:imageBase64]];
    VKRequest *req = [VKApi uploadAlbumPhotoRequest:image parameters:[VKImageParameters pngImage] albumId:album_id.integerValue groupId:group_id.integerValue];
    //VKRequest *req = [[VKApi photos] save:<#(NSDictionary *)#>];
    [self performRequest:req withCommand:command];
}

- (void)friends_get:(CDVInvokedUrlCommand*)command
{
    NSNumber *user_id = [command.arguments objectAtIndex:0];
    NSString *order = [command.arguments objectAtIndex:1];
    NSNumber *count = [command.arguments objectAtIndex:2];
    NSNumber *offset = [command.arguments objectAtIndex:3];
    NSString *fields = [command.arguments objectAtIndex:4];
    NSString *name_case = [command.arguments objectAtIndex:5];
    VKRequest *req = [[VKApi friends] get:@{VK_API_USER_ID: user_id, VK_API_ORDER: order, VK_API_COUNT: count, VK_API_OFFSET: offset, VK_API_FIELDS: fields, VK_API_NAME_CASE: name_case}];
    [self performRequest:req withCommand:command];
}

- (void)friends_getOnline:(CDVInvokedUrlCommand*)command
{
    NSNumber *user_id = [command.arguments objectAtIndex:0];
    NSString *order = [command.arguments objectAtIndex:1];
    NSNumber *count = [command.arguments objectAtIndex:2];
    NSNumber *offset = [command.arguments objectAtIndex:3];
    VKRequest *req = [VKRequest requestWithMethod:@"friends.getOnline" parameters:@{VK_API_USER_ID: user_id, VK_API_ORDER:order, VK_API_COUNT: count, VK_API_OFFSET: offset}];
    [self performRequest:req withCommand:command];
}

- (void)friends_getMutual:(CDVInvokedUrlCommand*)command
{
    NSNumber *source_uid = [command.arguments objectAtIndex:0];
    id target_uid = [command.arguments objectAtIndex:1];
    NSString *order = [command.arguments objectAtIndex:2];
    NSNumber *count = [command.arguments objectAtIndex:3];
    NSNumber *offset = [command.arguments objectAtIndex:4];
    NSMutableDictionary *params = [@{@"source_uid": source_uid, VK_API_ORDER: order, VK_API_COUNT: count, VK_API_OFFSET: offset} mutableCopy];
    if([target_uid isKindOfClass:NSNumber.class]) {
        params[@"target_uid"] = target_uid;
    } else if([target_uid isKindOfClass:NSString.class]) {
        params[@"target_uids"] = target_uid;
    }
    VKRequest *req = [VKRequest requestWithMethod:@"friends.getMutual" parameters:params];
    [self performRequest:req withCommand:command];
}

- (void)friends_getRecent:(CDVInvokedUrlCommand*)command
{
    NSNumber *count = [command.arguments objectAtIndex:0];
    VKRequest *req = [VKRequest requestWithMethod:@"friends.getRecent" parameters:@{VK_API_COUNT: count}];
    [self performRequest:req withCommand:command];
}

- (void)friends_getRequests:(CDVInvokedUrlCommand*)command
{
    NSNumber *offset = [command.arguments objectAtIndex:0];
    NSNumber *count = [command.arguments objectAtIndex:1];
    NSNumber *extended = [command.arguments objectAtIndex:2];
    NSNumber *needs_mutual = [command.arguments objectAtIndex:3];
    NSNumber *out_ = [command.arguments objectAtIndex:4];
    NSNumber *sort = [command.arguments objectAtIndex:5];
    NSNumber *suggested = [command.arguments objectAtIndex:6];
    VKRequest *req = [VKRequest requestWithMethod:@"friends.getRequests" parameters:@{VK_API_OFFSET: offset, VK_API_COUNT: count, VK_API_EXTENDED: extended, @"needs_mutual": needs_mutual, @"out": out_, VK_API_SORT: sort, @"suggested": suggested}];
    [self performRequest:req withCommand:command];
}

- (void)callApiMethod:(CDVInvokedUrlCommand *)command
{
    NSString *method = [command.arguments objectAtIndex:0];
    NSDictionary *params = [command.arguments objectAtIndex:1];
    VKRequest *req = [VKRequest requestWithMethod:method parameters:params];
    [self performRequest:req withCommand:command];
}

#pragma mark - VKSdkDelegate


- (void)vkSdkAccessAuthorizationFinishedWithResult:(VKAuthorizationResult *)result;
{
    NSLog(@"VK Access authorization finished");
    if(result.error) {
        NSLog(@"VK Error %@", result.error);
        if(vkCallBackBlock) {
            vkCallBackBlock(nil, result.error.description);
            vkCallBackBlock = nil;
        } else if(savedCommand) {
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:result.error.description];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:savedCommand.callbackId];
        }
    }
}

- (void)vkSdkUserAuthorizationFailed;
{
    NSLog(@"VK User authorization failed");
}

- (void)vkSdkAccessTokenUpdated:(VKAccessToken *)newToken oldToken:(VKAccessToken *)oldToken;
{
    NSLog(@"VK Token %@", newToken.accessToken);
    if(vkCallBackBlock) vkCallBackBlock(newToken.accessToken, nil);
    vkCallBackBlock = nil;
}

- (void)vkSdkTokenHasExpired:(VKAccessToken *)expiredToken;
{
    NSLog(@"VK Token has expired");
}


#pragma mark - VKSdkUIDelegate

- (void)vkSdkShouldPresentViewController:(UIViewController *)controller;
{
    [[self findViewController] presentViewController:controller animated:YES completion:nil];
}

- (void)vkSdkNeedCaptchaEnter:(VKError *)captchaError;
{
    NSLog(@"Need captcha %@", captchaError);
}

- (void)vkSdkWillDismissViewController:(UIViewController *)controller;
{
    NSLog(@"VK view controller will be dismissed");
}

- (void)vkSdkDidDismissViewController:(UIViewController *)controller;
{
    NSLog(@"VK view controller dismissed");
}


@end

File: src/windows/SocialVk/SocialVk.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json;
using VK.WindowsPhone.SDK;
using VK.WindowsPhone.SDK.API;
using VK.WindowsPhone.SDK.API.Model;
using VK.WindowsPhone.SDK.Util;
using VK.WindowsPhone.SDK.Pages;
using Social.Cordova.JSON;
using System.IO;
using System.Diagnostics;
using Windows.UI.Core;
using Windows.UI.Xaml;
using Windows.ApplicationModel.Core;
using Windows.ApplicationModel.Activation;

namespace Social
{
    public sealed class EventArgs {
        public int callbackid { get; set; }
        public string error { get; set; }
        public string result { get; set; }
    }

    public sealed class SocialVk
    {
        private static SocialVk instance;
        public event EventHandler<EventArgs> callback;
        private int lastCbId = 0;
        private CoreDispatcher dispatcher = CoreApplication.MainView.CoreWindow.Dispatcher;

        public SocialVk() {
            instance = this;
        }

        public static SocialVk Instance {
            get {
                return instance;
            }
        }

        private void sendResult(int cbid = 0, string result = "", string error = "") {
            if (cbid <= 0) cbid = lastCbId;
            if (cbid <= 0) {
                Debug.WriteLine("SocialVk: Wrong callback ID!");
                return;
            }
            if (callback != null) {
                if (result == null) result = "null";
                if (error == null) error = "null";
                callback(this, new EventArgs() { callbackid = cbid, result = result, error = error });
            } else
                Debug.WriteLine("SocialVk: Callback not defined!");
        }

        public void OnActivated(IActivatedEventArgs args) {
            try {
                if (args.Kind == ActivationKind.Protocol) {
                    var protocolArgs = args as ProtocolActivatedEventArgs;
                    if (protocolArgs != null) 
                        VK.WindowsPhone.SDK.VKProtocolActivationHelper.HandleProtocolLaunch(protocolArgs);
                    else {
                        var webProtocolArgs = args as Windows.UI.WebUI.WebUIProtocolActivatedEventArgs;
                        VK.WindowsPhone.SDK.VKProtocolActivationHelper.HandleWebProtocolLaunch(webProtocolArgs);
                    }
                }
            } catch (Exception e) {
                sendResult(0, "", e.ToString());
            }
        }

        public void init(string appId, int cbid) {

            VKSDK.AccessTokenReceived += (sender, arg) => {
                Debug.WriteLine("Access token recieved " + arg);
                JObject result = new JObject();
                result["token"] = arg.NewToken.AccessToken;
                VKRequest.Dispatch<List<VKUser>>(
                                new VKRequestParameters(
                                    "users.get",
                                    "fields", "id, nickname, first_name, last_name, sex, bdate, timezone, photo, photo_big, city, country"),
                                (res) => {
                                    if (res.ResultCode == VKResultCode.Succeeded) {
                                        string user = JsonConvert.SerializeObject(res.Data[0]);
                                        result["user"] = JObject.Parse(user);
                                        sendResult(0, result.ToString());
                                    } else {
                                        sendResult(0, "", res.ResultString);
                                    }
                                });
            };

            VKSDK.CaptchaRequest = (VKCaptchaUserRequest captchaUserRequest, Action<VKCaptchaUserResponse> action) => {
                Debug.WriteLine("Captcha request " + captchaUserRequest);
                sendResult(0, "", "Captcha request");
            };

            VKSDK.AccessDenied += (sender, arg) => {
                Debug.WriteLine("Access denied " + arg);
                sendResult(0, "", "Access denied");
            };

            VKSDK.Initialize(appId);
            VKSDK.WakeUpSession();

            Debug.WriteLine("VK plugin inited!");
            sendResult(cbid, "VK Plugin inited");
        }

        public void login(string par, int cbid) {
            try {
                lastCbId = cbid;
                string[] _scope = JsonHelper.Deserialize<string[]>(par);
                if (_scope == null) {
                    Debug.WriteLine("Empty permissions list");
                    sendResult(cbid, "", "Empty permissions list." + par);
                } else {
                    List<String> scope = new List<string>(_scope);
                    //Window.Current.Dispatcher.RunAsync(CoreDispatcherPriority.Normal,
                    //Windows.ApplicationModel.Core.CoreApplication.MainView.CoreWindow.Dispatcher.RunAsync(CoreDispatcherPriority.Normal,
                    //    () =>
                    //    {
                            try {
                                string err = VKSDK.Authorize(scope, false, false, LoginType.VKApp);
                                if (err != null)
                                    sendResult(cbid, "", err);
                            } catch (Exception e) {
                                sendResult(cbid, "", e.ToString());
                            }
                    //    }
                    //);
                }
            } catch (Exception e) {
                sendResult(cbid, "", e.ToString() + e.StackTrace + e.Message + e.Source);
            }
        }

        public void share(string sourceUrl, string description, string imageUrl, int cbid) {
            sendResult(cbid, "", "Not implemented");
        }

        public void logout(int cbid) {
            VKSDK.Logout();
            sendResult(cbid, "Logout completed");
        }

        public void users_get(string user_ids, string fields, string name_case, int cbid) {
            VKRequest.Dispatch<List<VKUser>>(
                            new VKRequestParameters(
                                "users.get",
                                "user_ids", user_ids,
                                "fields", fields),
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    //var user = res.Data[0];
                                    //DispatchCommandResult(new PluginResult(PluginResult.Status.OK, res.ResultString), options.Last<string>());
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    //DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR, res.ResultString), options.Last<string>());
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void users_search(string query_or_params, int cbid) {
            VKRequestParameters vkrp;
            try {
                JObject p = JObject.Parse(query_or_params);
                var dict = p.ToObject<Dictionary<string, string>>();
                vkrp = new VKRequestParameters("users.search", dict);
            } catch (Exception e) {
                vkrp = new VKRequestParameters("users.search", "q", query_or_params);
            }
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void users_isAppUser(string user_id, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("users.isAppUser", "user_id", user_id);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            },
                            (json) => new Object());
        }

        public void users_getSubscriptions(string user_id, string extended, string offset, string count, string fields, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("users.getSubscriptions", "user_id", user_id, "extended", extended, "offset", offset, "count", count, "fields", fields);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void users_getFollowers(string user_id, string offset, string count, string fields, string name_case, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("users.getFollowers", "user_id", user_id, "offset", offset, "count", count, "fields", fields, "name_case", name_case);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void wall_post(string par) {
            /*
            string[] options = JsonHelper.Deserialize<string[]>(par);
            VKPublishInputData data = new VKPublishInputData();
            try {
                JObject p = JObject.Parse(options[0]);
                var dict = p.ToObject<Dictionary<string, string>>();
                data.Text = dict["message"];
                // TODO links and images
            } catch (Exception e) {
                data.Text = options[0];
            }
            VKSDK.Publish(data);
            */
        }

        public void photos_getUploadServer(string album_id, string group_id, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("photos.getUploadServer", "album_id", album_id, "group_id", group_id);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void photos_getWallUploadServer(string group_id, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("photos.getWallUploadServer", "group_id", group_id);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void photos_saveWallPhoto(string imageBase64, string user_id, string group_id, int cbid) {
            long userId = long.Parse(user_id);
            long groupdId = long.Parse(group_id);
            byte[] bytes;
            int comma = imageBase64.IndexOf(',');
            if (comma >= 0) {
                bytes = Convert.FromBase64String(imageBase64.Substring(comma + 1));
            } else {
                bytes = Convert.FromBase64String(imageBase64);
            }
            var ms = new MemoryStream(bytes, 0, bytes.Length);
            VKUploadRequest req = VKUploadRequest.CreatePhotoWallUploadRequest(userId);
            req.Dispatch(ms, (progress) => {
            }, (res) => {
                if (res.ResultCode == VKResultCode.Succeeded) {
                    sendResult(cbid, res.ResultString);
                } else {
                    string errtext;
                    if (res.ResultString != null) errtext = res.ResultString;
                    else if (res.Error != null) errtext = res.Error.error_msg;
                    else errtext = "Unknown error";
                    sendResult(cbid, "", errtext);
                }
            });
        }

        public void photos_save(string imageBase64, string album_id, string group_id, int cbid) {
            long albumId = long.Parse(album_id);
            long groupdId = long.Parse(group_id);
            byte[] bytes;
            int comma = imageBase64.IndexOf(',');
            if (comma >= 0) {
                bytes = Convert.FromBase64String(imageBase64.Substring(comma + 1));
            } else {
                bytes = Convert.FromBase64String(imageBase64);
            }
            var ms = new MemoryStream(bytes, 0, bytes.Length);
            VKUploadRequest req = VKUploadRequest.CreatePhotoAlbumUploadRequest(albumId, groupdId);
            req.Dispatch(ms, (progress) => {
            }, (res) => {
                if (res.ResultCode == VKResultCode.Succeeded) {
                    sendResult(cbid, res.ResultString);
                } else {
                    string errtext;
                    if (res.ResultString != null) errtext = res.ResultString;
                    else if (res.Error != null) errtext = res.Error.error_msg;
                    else errtext = "Unknown error";
                    sendResult(cbid, "", errtext);
                }
            });
        }

        public void friends_get(string user_id, string order, string count, string offset, string fields, string name_case, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("friends.get", "user_id", user_id, "order", order, "count", count, "offset", offset, "fields", fields, "name_case", name_case);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void friends_getOnline(string user_id, string order, string count, string offset, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("friends.getOnline", "user_id", user_id, "order", order, "count", count, "offset", offset);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void friends_getMutual(string source_uid, string target_uid, string order, string count, string offset, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("friends.getMutual", "source_uid", source_uid, "target_uid", target_uid, "order", order, "count", count, "offset", offset);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void friends_getRecent(string count, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("friends.getRecent", "count", count);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void friends_getRequests(string offset, string count, string extended, string needs_mutual, string _out, string sort, string suggested, int cbid) {
            VKRequestParameters vkrp;
            vkrp = new VKRequestParameters("friends.getRequests", "offset", offset, "count", count, "extended", extended, "needs_mutual", needs_mutual, "out", _out, "sort", sort, "suggested", suggested);
            VKRequest.Dispatch<object>(
                            vkrp,
                            (res) => {
                                if (res.ResultCode == VKResultCode.Succeeded) {
                                    sendResult(cbid, res.ResultString);
                                } else {
                                    sendResult(cbid, "", res.ResultString);
                                }
                            });
        }

        public void callApiMethod(string method, string args, int cbid) {
            VKRequestParameters vkrp = null;
            try {
                JObject p = JObject.Parse(args);
                var dict = p.ToObject<Dictionary<string, string>>();
                vkrp = new VKRequestParameters(method, dict);
            } catch (Exception e) {

            }
            if (vkrp != null)
                VKRequest.Dispatch<object>(
                                vkrp,
                                (res) => {
                                    if (res.ResultCode == VKResultCode.Succeeded) {
                                        sendResult(cbid, res.ResultString);
                                    } else {
                                        sendResult(cbid, "", res.ResultString);
                                    }
                                });
            else
                sendResult(cbid, "", "Invalid parameters: " + method + " " + args);
        }
    }
}
